modules/phase_field/src/materials/KKSPhaseConcentrationMultiPhaseMaterial.C

```diff
@@ -0,0 +1,260 @@
+    _iter(declareProperty<Real>("nested_iterations")),
+    _abs_tol(getParam<Real>("absolute_tolerance")),
+    _rel_tol(getParam<Real>("relative_tolerance")),
+    _nested_solve(NestedSolve(parameters))
+
+{
+  for (unsigned int m = 0; m < _num_c * _num_j; ++m)
+  {
+    _ci_old[m] = &getMaterialPropertyOld<Real>(_ci_names[m]);
+    _prop_ci[m] = &declareProperty<Real>(_ci_names[m]);
+  }
+
+  for (unsigned int m = 0; m < _num_j; ++m)
+  {
+    _prop_Fi[m] = &getMaterialPropertyByName<Real>(_Fj_names[m]);
+    _Fi_copy[m] = &declareProperty<Real>("cp" + _Fj_names[m]);
+  }
+
+  for (unsigned int m = 0; m < _num_j; ++m)
+  {
+    _prop_hj[m] = &getMaterialPropertyByName<Real>(_hj_names[m]);
+
+    _dFidci[m].resize(_num_c);
+    _dFidci_copy[m].resize(_num_c);
+    _d2Fidcidbi[m].resize(_num_c);
+    _d2Fidcidbi_copy[m].resize(_num_c);
+
+    for (unsigned int n = 0; n < _num_c; ++n)
+    {
+      _dFidci[m][n] = &getMaterialPropertyDerivative<Real>(_Fj_names[m], _ci_names[m + n * _num_j]);
+      _dFidci_copy[m][n] =
+          &declarePropertyDerivative<Real>("cp" + _Fj_names[m], _ci_names[m + n * _num_j]);
+
+      _d2Fidcidbi[m][n].resize(_num_c);
+      _d2Fidcidbi_copy[m][n].resize(_num_c);
+
+      for (unsigned int l = 0; l < _num_c; ++l)
+      {
+        _d2Fidcidbi[m][n][l] = &getMaterialPropertyDerivative<Real>(
+            _Fj_names[m], _ci_names[m + n * _num_j], _ci_names[m + l * _num_j]);
+        _d2Fidcidbi_copy[m][n][l] = &declarePropertyDerivative<Real>(
+            "cp" + _Fj_names[m], _ci_names[m + n * _num_j], _ci_names[m + l * _num_j]);
+      }
+    }
+  }
+
+  // partial derivative of Fi wrt coupled variables, to be passed to kernels
+  for (unsigned int m = 0; m < _num_j; ++m)
+  {
+    _dFidarg[m].resize(_n_args);
+    _dFidarg_copy[m].resize(_n_args);
+
+    for (unsigned int n = 0; n < _n_args; ++n)
+    {
+      _dFidarg[m][n] = &getMaterialPropertyDerivative<Real>(_Fj_names[m], _args_names[n]);
+      _dFidarg_copy[m][n] = &declarePropertyDerivative<Real>("cp" + _Fj_names[m], _args_names[n]);
+    }
+  }
+
+  // second partial derivatives of F1 wrt c1 and another coupled variable, to be passed to kernels
+  for (unsigned int m = 0; m < _num_c; ++m)
+  {
+    _d2F1dc1darg[m].resize(_n_args);
+    _d2F1dc1darg_copy[m].resize(_n_args);
+
+    for (unsigned int n = 0; n < _n_args; ++n)
+    {
+      _d2F1dc1darg[m][n] =
+          &getMaterialPropertyDerivative<Real>(_Fj_names[0], _ci_names[m * _num_j], _args_names[n]);
+      _d2F1dc1darg_copy[m][n] = &declarePropertyDerivative<Real>(
+          "cp" + _Fj_names[0], _ci_names[m * _num_j], _args_names[n]);
+    }
+  }
+}
+
+void
+KKSPhaseConcentrationMultiPhaseMaterial::initQpStatefulProperties()
+{
+  for (unsigned int m = 0; m < _num_c * _num_j; ++m)
+    (*_prop_ci[m])[_qp] = _ci_IC[m];
+}
+
+void
+KKSPhaseConcentrationMultiPhaseMaterial::initialSetup()
+{
+  for (unsigned int m = 0; m < _num_j; ++m)
+    _Fj_mat[m] = &getMaterialByName(_Fj_names[m]);
+}
```

Comment:
@jiangwen84 I have a question on this piece of code. `_Fj_mat` is a vector of MaterialBase pointers. How to properly initialize it from "Fj_material"? (I did not have to do this in the two-phase case because I initialized the two free energy MaterialBase separately, so no vector in two-phase case.) This current version can compile but received segmentation fault here when running an input file.